src/sentry/sentry_metrics/extraction_rules.py

```diff
@@ -0,0 +1,114 @@
+from collections.abc import Mapping, Sequence
+from dataclasses import dataclass
+from typing import Any
+
+from sentry.models.project import Project
+from sentry.utils import json
+
+METRICS_EXTRACTION_RULES_OPTION_KEY = "sentry:metrics_extraction_rules"
+
+
+class MetricsExtractionRuleValidationError(ValueError):
+    pass
+
+
+@dataclass(frozen=True)
+class MetricsExtractionRule:
+    span_attribute: str
+    type: str
+    unit: str
+    tags: set[str]
+
+    @classmethod
+    def from_dict(cls, dictionary: Mapping[str, Any]) -> "MetricsExtractionRule":
+        return MetricsExtractionRule(
+            span_attribute=dictionary["span_attribute"],
+            type=dictionary["type"],
+            unit=dictionary["unit"],
+            tags=set(dictionary.get("tags") or set()),
+        )
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return self.__dict__
+
+    def generate_mri(self, use_case: str = "custom"):
```

Comment:
nit: maybe validate the use case here using `use_case_utils.string_to_use_case_id`

Reply:
nit: also since `generate_mri` is called in both update and delete, it might be easier to have it as a field